web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts

```diff
@@ -194,6 +197,48 @@ export default class TtyTerminal {
     this.term.focus();
   }
 
+  loadSearchAddon({
+    onDidChangeResults,
+  }: {
+    onDidChangeResults: (resultIndex: number, resultCount: number) => void;
+  }) {
+    this.searchAddon.onDidChangeResults(ev =>
+      onDidChangeResults(ev.resultIndex, ev.resultCount)
+    );
+    this.term.loadAddon(this.searchAddon);
+  }
+
+  search(
+    searchString: string,
+    searchPrevious: boolean,
+    searchOpts?: ISearchOptions
+  ) {
+    if (!this.searchAddon) {
+      throw new Error(
+        '_searchAddon has not been loaded. Use `loadSearchAddon` before searching.'
+      );
+    }
+
+    const opts: ISearchOptions = {
+      // TODO (avatus): we are setting these defaults (regex/caseSensitive) to
+      // what we think would be best UX, but the search UI could be expanded upon
+      // to allow these values to be updated.
+      regex: true,
+      caseSensitive: false,
+      ...searchOpts,
+    };
+
+    if (searchPrevious) {
+      this.searchAddon.findPrevious(searchString, opts);
+    } else {
+      this.searchAddon.findNext(searchString, opts);
+    }
+  }
+
+  clearSearch() {
+    this.searchAddon.clearDecorations();
+  }
+
```

Comment:
We have the same logic in Web UI and Connect. It seems to me that we don't need this "proxy code" and the `TerminalSearch` could just directly operate on the `searchAddon`. The callsite would only have to load the addon and then pass it to `TerminalSearch`.
Here is what I mean (I only updated Connect):

<details>

<summary>Patch</summary>

```
Subject: [PATCH] Pass search addon to `TerminalSearch`
---
Index: web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx b/web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx
--- a/web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx	(revision 93d9d74244e88660318ce22dd98d1d4ed2769226)
+++ b/web/packages/teleterm/src/ui/DocumentTerminal/DocumentTerminal.tsx	(date 1731412147824)
@@ -149,7 +149,8 @@
                   show={showSearch}
                   onClose={onSearchClose}
                   onOpen={onSearchOpen}
-                  xTerm={terminalRef.current}
+                  searchAddon={terminalRef.current.getSearchAddon()}
+                  focusTerminal={terminalRef.current.focus}
                 />
                 {$fileTransfer}
               </>
Index: web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts b/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts
--- a/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts	(revision 93d9d74244e88660318ce22dd98d1d4ed2769226)
+++ b/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/ctrl.ts	(date 1731411359982)
@@ -19,7 +19,7 @@
 import '@xterm/xterm/css/xterm.css';
 import { IDisposable, ITheme, Terminal } from '@xterm/xterm';
 import { FitAddon } from '@xterm/addon-fit';
-import { SearchAddon, ISearchOptions } from 'shared/components/TerminalSearch';
+import { SearchAddon } from 'shared/components/TerminalSearch';
 import { debounce } from 'shared/utils/highbar';
 
 import { WindowsPty } from 'teleterm/services/pty';
@@ -171,6 +171,7 @@
     });
 
     this.fitAddon.fit();
+    this.term.loadAddon(this.searchAddon);
 
     this.term.onData(data => {
       this.ptyProcess.write(data);
@@ -197,46 +198,8 @@
     this.term.focus();
   }
 
-  loadSearchAddon({
-    onDidChangeResults,
-  }: {
-    onDidChangeResults: (resultIndex: number, resultCount: number) => void;
-  }) {
-    this.searchAddon.onDidChangeResults(ev =>
-      onDidChangeResults(ev.resultIndex, ev.resultCount)
-    );
-    this.term.loadAddon(this.searchAddon);
-  }
-
-  search(
-    searchString: string,
-    searchPrevious: boolean,
-    searchOpts?: ISearchOptions
-  ) {
-    if (!this.searchAddon) {
-      throw new Error(
-        '_searchAddon has not been loaded. Use `loadSearchAddon` before searching.'
-      );
-    }
-
-    const opts: ISearchOptions = {
-      // TODO (avatus): we are setting these defaults (regex/caseSensitive) to
-      // what we think would be best UX, but the search UI could be expanded upon
-      // to allow these values to be updated.
-      regex: true,
-      caseSensitive: false,
-      ...searchOpts,
-    };
-
-    if (searchPrevious) {
-      this.searchAddon.findPrevious(searchString, opts);
-    } else {
-      this.searchAddon.findNext(searchString, opts);
-    }
-  }
-
-  clearSearch() {
-    this.searchAddon.clearDecorations();
+  getSearchAddon() {
+    return this.searchAddon;
   }
 
   requestResize(): void {
Index: web/packages/shared/components/TerminalSearch/TerminalSearch.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/packages/shared/components/TerminalSearch/TerminalSearch.tsx b/web/packages/shared/components/TerminalSearch/TerminalSearch.tsx
--- a/web/packages/shared/components/TerminalSearch/TerminalSearch.tsx	(revision 93d9d74244e88660318ce22dd98d1d4ed2769226)
+++ b/web/packages/shared/components/TerminalSearch/TerminalSearch.tsx	(date 1731412147842)
@@ -19,20 +19,20 @@
 import { useEffect, useState, useRef, useCallback } from 'react';
 import styled, { useTheme } from 'styled-components';
 import { Flex, ButtonIcon, Box, P2, Input } from 'design';
-import { Terminal as Xterm } from '@xterm/xterm';
-import { ISearchOptions } from '@xterm/addon-search';
 import { ArrowDown, ArrowUp, Cross } from 'design/Icon';
 
 export const TerminalSearch = ({
-  xTerm,
   show,
   onClose,
   onOpen,
+  searchAddon,
+  focusTerminal,
 }: {
-  xTerm: TerminalWithSearch;
   show: boolean;
   onClose(): void;
   onOpen(): void;
+  focusTerminal(): void;
+  searchAddon: SearchAddon;
 }) => {
   const theme = useTheme();
   const searchInputRef = useRef<HTMLInputElement>();
@@ -43,60 +43,66 @@
     resultCount: number;
   }>({ resultIndex: 0, resultCount: 0 });
 
-  function search(value: string, prev?: boolean) {
-    if (!xTerm) {
-      return;
-    }
+  function search(value: string, direction: 'next' | 'previous') {
     const match = theme.colors.terminal.brightYellow;
     const activeMatch = theme.colors.terminal.yellow;
     setSearchValue(value);
 
-    xTerm.search(value, prev, {
+    const options = {
       decorations: {
         matchOverviewRuler: match,
         activeMatchColorOverviewRuler: activeMatch,
         matchBackground: match,
         activeMatchBackground: activeMatch,
       },
-    });
+    };
+
+    if (direction === 'next') {
+      searchAddon.findNext(value, options);
+    } else {
+      searchAddon.findPrevious(value, options);
+    }
   }
 
   const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
-    search(e.target.value);
+    search(e.target.value, 'next');
   };
 
   const onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
     if (e.key === 'Enter') {
-      search(e.currentTarget.value);
+      search(e.currentTarget.value, 'next');
     }
   };
 
   function searchNext() {
-    search(searchValue);
+    search(searchValue, 'next');
   }
 
   function searchPrevious() {
-    search(searchValue, true /* search for previous result */);
+    search(searchValue, 'previous');
   }
 
   const onEscape = useCallback(() => {
     onClose();
-    xTerm.clearSearch();
-    xTerm.term.focus();
-  }, [xTerm, onClose]);
+    searchAddon.clearDecorations();
+    focusTerminal();
+  }, [focusTerminal, onClose, searchAddon]);
 
   const onSearchResultsChange = useCallback(
-    (resultIndex: number, resultCount: number) => {
+    ({
+      resultIndex,
+      resultCount,
+    }: {
+      resultIndex: number;
+      resultCount: number;
+    }) => {
       setSearchResults({ resultIndex, resultCount });
     },
     []
   );
 
   useEffect(() => {
-    if (!xTerm) {
-      return;
-    }
-    xTerm.loadSearchAddon({ onDidChangeResults: onSearchResultsChange });
+    searchAddon.onDidChangeResults(onSearchResultsChange);
     const listener = (e: KeyboardEvent) => {
       if (e.key === 'Escape') {
         onEscape();
@@ -114,7 +120,7 @@
     return () => {
       document.removeEventListener('keydown', listener);
     };
-  }, [onEscape, onOpen, searchInputRef, onSearchResultsChange, xTerm]);
+  }, [onEscape, onOpen, searchInputRef, onSearchResultsChange, searchAddon]);
 
   if (!show) {
     return;
@@ -166,19 +172,37 @@
   background-color: ${p => p.theme.colors.levels.popout};
 `;
 
-export interface TerminalWithSearch {
-  term: Xterm;
-  search(
-    searchString: string,
-    searchPrevious: boolean,
-    searchOpts?: ISearchOptions
+interface SearchAddon {
+  clearDecorations(): void;
+  onDidChangeResults(
+    onSearchResultsChange: ({
+      resultIndex,
+      resultCount,
+    }: {
+      resultIndex: number;
+      resultCount: number;
+    }) => void
   ): void;
-  clearSearch(): void;
-  // loading this addon must be done by the caller so it can provide its 'onDidChangeResults' callback.
-  // This can be done after the addon is loaded, but then we'd still need the callback in the props anyway
-  loadSearchAddon({
-    onDidChangeResults,
-  }: {
-    onDidChangeResults: (resultIndex: number, resultCount: number) => void;
-  }): void;
+  findNext(
+    value: string,
+    options: {
+      decorations: {
+        activeMatchColorOverviewRuler: string;
+        matchOverviewRuler: string;
+        activeMatchBackground: string;
+        matchBackground: string;
+      };
+    }
+  ): void;
+  findPrevious(
+    value: string,
+    options: {
+      decorations: {
+        activeMatchColorOverviewRuler: string;
+        matchOverviewRuler: string;
+        activeMatchBackground: string;
+        matchBackground: string;
+      };
+    }
+  ): void;
 }
Index: web/packages/teleterm/src/ui/DocumentTerminal/Terminal/Terminal.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/Terminal.tsx b/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/Terminal.tsx
--- a/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/Terminal.tsx	(revision 93d9d74244e88660318ce22dd98d1d4ed2769226)
+++ b/web/packages/teleterm/src/ui/DocumentTerminal/Terminal/Terminal.tsx	(date 1731412147835)
@@ -26,7 +26,6 @@
   makeErrorAttemptWithStatusText,
   makeSuccessAttempt,
 } from 'shared/hooks/useAsync';
-import { TerminalWithSearch } from 'shared/components/TerminalSearch/TerminalSearch';
 
 import { WindowsPty } from 'teleterm/services/pty';
 import { IPtyProcess } from 'teleterm/sharedProcess/ptyHost';
@@ -59,7 +58,7 @@
   keyboardShortcutsService: KeyboardShortcutsService;
   // terminalAddons is used to pass the tty to the parent component to enable any optional components like search or filetransfers.
   terminalAddons?: (
-    terminalRef: React.MutableRefObject<TerminalWithSearch>
+    terminalRef: React.MutableRefObject<XTermCtrl>
   ) => React.JSX.Element;
 };
 ```

</details>

Reply:
I couldn't use the patch (for some reason anytime i try to apply a patch it says corrupted on last line).
at any rate, i think i took care of what you were talking about. let me know

Reply:
Oops. I generated it with GoLand and I can still apply it there. But with CLI it fails, I think I need to generate it with CLI as well.